File: Source/Fantasy.Jobs/UntilSuccess.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.XSerialization;
using System.Threading;
using Fantasy.ServiceModel;

namespace Fantasy.Jobs
{

    [Instruction]
    [XSerializable("until-success", NamespaceUri = Consts.XNamespaceURI)]  
    internal class UntilSuccess : AbstractInstruction
    {

        [XAttribute("failIfAllSkipped")]
        private bool _failIfAllSkipped = true;

       

        public override void Execute()
        {
            if (this._items.Count > 0)
            {
                IConditionService conditionSvc = this.Site.GetRequiredService<IConditionService>();
                IJob job = this.Site.GetRequiredService<IJob>();
                ILogger logger = this.Site.GetService<ILogger>();
                int index = job.RuntimeStatus.Local.GetValue("until-success.index", 0);
                bool success = false;
                bool hasException = false;
                while (index < this._items.Count && !success)
                {
                    try
                    {
                        Try chance = this._items[index];
                        if (conditionSvc.Evaluate(chance))
                        {
                            if (logger != null)
                            {
                                logger.LogMessage(LogCategories.Instruction, "Execute try No.{0}", index);
                            }
                            job.ExecuteInstruction(chance);
                            success = true;
                        }
                        else
                        {
                            logger.LogMessage(LogCategories.Instruction, "Skip try No.{0}", index);
                        }
                        
                    }
                    catch (ThreadAbortException)
                    {

                    }
                    catch (Exception error)
                    {
                        hasException = true;
                        job.RuntimeStatus.Local["until-success.index"] = index;
                        if (logger != null)
                        {
                            logger.LogError(LogCategories.Instruction, error,  "try instruction faild, try next one.");

                        }
                    }
                    index++;
                }

               
                if (!success)
                {
                   

                    if (hasException)
                    {
                        if (logger != null)
                        {
                            logger.LogError(LogCategories.Instruction, Properties.Resources.UnitlSuccessFailedText);
                        }
                        throw new JobException(Properties.Resources.UnitlSuccessFailedText);
                    }
                    else
                    {
                        if (logger != null)
                        {
                            logger.LogError(LogCategories.Instruction, Properties.Resources.UntilSuccessAllSkippedText);
                            
                        }
                        if (this._failIfAllSkipped)
                        {
                            throw new JobException(Properties.Resources.UntilSuccessAllSkippedText);
                        }
                    }
                }
               
            }
        }
        [XArray(Order = 10)]
        [XArrayItem(Name = "try", Type = typeof(Try))]
        private IList<Try> _items = new List<Try>();

       
    }
}

File: Fantasy.Jobs/src/fantasy/jobs/UntilSuccess.java
package fantasy.jobs;

import fantasy.xserialization.*;
import fantasy.servicemodel.*;
import fantasy.jobs.properties.*;

@Instruction
@XSerializable(name = "until-success", namespaceUri = Consts.XNamespaceURI)
public class UntilSuccess extends AbstractInstruction
{
	@XAttribute(name = "failIfAllSkipped")
	private boolean _failIfAllSkipped = true;



	@Override
	public void Execute() throws Exception
	{
		if (this._items.size() > 0)
		{
			IConditionService conditionSvc = this.getSite().getRequiredService(IConditionService.class);
			IJob job = this.getSite().getRequiredService(IJob.class);
			ILogger logger = this.getSite().getService(ILogger.class);
			int index = job.getRuntimeStatus().getLocal().GetValue("until-success.index", 0);
			boolean success = false;
			boolean hasException = false;
			while (index < this._items.size() && !success)
			{
				if(Thread.interrupted())
				{
					throw new InterruptedException();
				}
				try
				{
					Try chance = this._items.get(index);
					if (conditionSvc.Evaluate(chance))
					{
						if (logger != null)
						{
							logger.LogMessage(LogCategories.getInstruction(), "Execute try No.{0}", index);
						}
						job.ExecuteInstruction(chance);
						success = true;
					}
					else
					{
						Log.SafeLogMessage(logger, LogCategories.getInstruction(), "Skip try No.{0}", index);
					}

				}
				catch (InterruptedException e)
				{
                     throw e;
				}
				catch (RuntimeException error)
				{
					hasException = true;
					job.getRuntimeStatus().getLocal().setItem("until-success.index", index);
					if (logger != null)
					{
						logger.LogError(LogCategories.getInstruction(), error, "try instruction faild, try next one.");

					}
				}
				index++;
			}


			if (!success)
			{


				if (hasException)
				{
					if (logger != null)
					{
						logger.LogError(LogCategories.getInstruction(), Resources.getUnitlSuccessFailedText());
					}
					throw new JobException(Resources.getUnitlSuccessFailedText());
				}
				else
				{
					if (logger != null)
					{
						logger.LogError(LogCategories.getInstruction(), Resources.getUntilSuccessAllSkippedText());

					}
					throw new JobException(Resources.getUntilSuccessAllSkippedText());
				}
			}

		}
	}

	
	
	@XArray(order = 10, items=@XArrayItem(name = "try", type = Try.class))
	private java.util.List<Try> _items = new java.util.ArrayList<Try>();


}